src/internal/packager/images/push.go

```diff
@@ -46,85 +41,95 @@ func (i *ImageConfig) PushToZarfRegistry() error {
 	}
 
 	// If this is not a no checksum image push we will be pushing two images (the second will go faster as it checks the same layers)
-	if !i.NoChecksum {
+	if !cfg.NoChecksum {
 		totalSize = totalSize * 2
 	}
 
-	httpTransport := http.DefaultTransport.(*http.Transport).Clone()
-	httpTransport.TLSClientConfig.InsecureSkipVerify = i.Insecure
-	// TODO (@WSTARR) This is set to match the TLSHandshakeTimeout to potentially mitigate effects of https://github.com/defenseunicorns/zarf/issues/1444
-	httpTransport.ResponseHeaderTimeout = 10 * time.Second
-	progressBar := message.NewProgressBar(totalSize, fmt.Sprintf("Pushing %d images to the zarf registry", len(i.ImageList)))
-	defer progressBar.Stop()
-	craneTransport := helpers.NewTransport(httpTransport, progressBar)
-
-	pushOptions := config.GetCraneOptions(i.Insecure, i.Architectures...)
-	pushOptions = append(pushOptions, config.GetCraneAuthOption(i.RegInfo.PushUsername, i.RegInfo.PushPassword))
-	pushOptions = append(pushOptions, crane.WithTransport(craneTransport))
-
 	var (
 		err         error
 		tunnel      *k8s.Tunnel
-		registryURL string
+		registryURL = cfg.RegInfo.Address
 	)
 
-	registryURL = i.RegInfo.Address
+	progress := message.NewProgressBar(totalSize, fmt.Sprintf("Pushing %d images", len(toPush)))
+	defer progress.Stop()
 
-	c, _ := cluster.NewCluster()
-	if c != nil {
-		registryURL, tunnel, err = c.ConnectToZarfRegistryEndpoint(i.RegInfo)
-		if err != nil {
-			return err
+	if err := helpers.Retry(func() error {
+		c, _ := cluster.NewCluster()
+		if c != nil {
+			registryURL, tunnel, err = c.ConnectToZarfRegistryEndpoint(cfg.RegInfo)
+			if err != nil {
+				return err
+			}
+			if tunnel != nil {
+				defer tunnel.Close()
+			}
 		}
-	}
 
-	if tunnel != nil {
-		defer tunnel.Close()
-	}
+		progress = message.NewProgressBar(totalSize, fmt.Sprintf("Pushing %d images", len(toPush)))
+		pushOptions := createPushOpts(cfg, progress)
 
-	pushImage := func(img v1.Image, name string) error {
-		if tunnel != nil {
-			return tunnel.Wrap(func() error { return crane.Push(img, name, pushOptions...) })
-		}
+		pushImage := func(img v1.Image, name string) error {
+			if tunnel != nil {
+				return tunnel.Wrap(func() error { return crane.Push(img, name, pushOptions...) })
+			}
 
-		return crane.Push(img, name, pushOptions...)
-	}
+			return crane.Push(img, name, pushOptions...)
+		}
 
-	for refInfo, img := range refInfoToImage {
-		refTruncated := helpers.Truncate(refInfo.Reference, 55, true)
-		progressBar.UpdateTitle(fmt.Sprintf("Pushing %s", refTruncated))
+		pushed := []transform.Image{}
+		defer func() {
+			for _, refInfo := range pushed {
+				delete(toPush, refInfo)
+			}
+		}()
```

Comment:
Seems like we don't need to manually cleanup resources here. The garbage collector will clean this up for us

Reply:
If I understand the code correctly, this is necessary because the code here is run in a helpers.retry loop. If there's a failed attempt we only want to retry on images we haven't pushed yet

Reply:
correct, no matter what we want to retry incrementally, as failure can happen partway through, there is no need to re-attempt images that have already succeeded

Reply:
That makes sense. Thanks for the explanation @AustinAbro321 @Noxsios